prog/vm/github_runner.rb

```diff
@@ -0,0 +1,162 @@
+        enable_ip4: true,
+        storage_encrypted: false
+      )
+
+      Sshable.create(
+        unix_user: "runner",
+        host: "temp_#{vm_st.id}",
+        raw_private_key_1: ssh_key.keypair
+      ) { _1.id = vm_st.id }
+
+      github_runner = GithubRunner.create(
+        installation_id: installation.id,
+        repository_name: repository_name,
+        label: label,
+        vm_id: vm_st.id
+      ) { _1.id = ubid.to_uuid }
+
+      Strand.create(prog: "Vm::GithubRunner", label: "start") { _1.id = github_runner.id }
+    end
+  end
+
+  def vm
+    @vm ||= github_runner.vm
+  end
+
+  def github_client
+    @github_client ||= Github.installation_client(github_runner.installation.installation_id)
+  end
+
+  def before_run
+    when_destroy_set? do
+      if strand.label != "destroy"
+        hop_destroy
+      end
+    end
+  end
+
+  label def start
+    register_deadline(:wait, 10 * 60)
+
+    hop_wait_vm
+  end
+
+  label def wait_vm
+    nap 5 unless vm.strand.label == "wait"
+    vm.sshable.update(host: vm.ephemeral_net4 || vm.ephemeral_net6&.nth(2).to_s)
+    hop_bootstrap_rhizome
+  end
+
+  label def bootstrap_rhizome
+    bud Prog::BootstrapRhizome, {"target_folder" => "common", "subject_id" => vm.id, "user" => "runner"}
+    hop_wait_bootstrap_rhizome
+  end
+
+  label def wait_bootstrap_rhizome
+    reap
+    hop_install_actions_runner if leaf?
+    donate
+  end
+
+  # TODO: Move this to golden image too
+  label def install_actions_runner
+    vm.sshable.cmd("curl -o actions-runner-linux-x64-2.308.0.tar.gz -L https://github.com/actions/runner/releases/download/v2.308.0/actions-runner-linux-x64-2.308.0.tar.gz")
+    vm.sshable.cmd("echo '9f994158d49c5af39f57a65bf1438cbae4968aec1e4fec132dd7992ad57c74fa  actions-runner-linux-x64-2.308.0.tar.gz' | shasum -a 256 -c")
+    vm.sshable.cmd("tar xzf ./actions-runner-linux-x64-2.308.0.tar.gz")
+
+    hop_register_runner
+  end
+
+  label def register_runner
+    unless github_runner.runner_id
+      # We use generate-jitconfig instead of registration-token because it's
+      # recommended by GitHub for security reasons.
+      # https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-just-in-time-runners
+      data = {name: github_runner.ubid.to_s, labels: ["ubicloud"], runner_group_id: 1}
+      response = github_client.post("/repos/#{github_runner.repository_name}/actions/runners/generate-jitconfig", data)
+      github_runner.update(runner_id: response[:runner][:id], ready_at: Time.now)
+      vm.sshable.cmd("common/bin/daemonizer 'sudo -u runner /home/runner/run.sh --jitconfig #{response[:encoded_jit_config].shellescape}' runner-script")
+    end
+
+    case vm.sshable.cmd("common/bin/daemonizer --check runner-script")
+    when "Succeeded", "InProgress"
+      hop_wait
+    when "NotStarted", "Failed"
+      github_client.delete("/repos/#{github_runner.repository_name}/actions/runners/#{github_runner.runner_id}")
```

Comment:
Why do we delete the runner if it is not started?

Reply:
Actually, looking at the code, --check can never return NotStarted because we always start the command in the above block.

Reply:
Removed it.